server.nix

```diff
@@ -33,27 +38,83 @@
 
             # dontCheck: https://github.com/gleachkr/Carnap/issues/123
-            Carnap        = dontCheck (oldpkgs.callPackage ./Carnap/Carnap.nix { });
-            Carnap-Client = oldpkgs.callPackage ./Carnap-Client/Carnap-Client.nix { };
+            Carnap        = withGitignore
+                (dontCheck (oldpkgs.callPackage ./Carnap/Carnap.nix { }));
+            Carnap-Client = withGitignore
+                (oldpkgs.callPackage ./Carnap-Client/Carnap-Client.nix { });
+
             # for client development with ghc
-            # Carnap-GHCJS  = oldpkgs.callPackage ./Carnap-GHCJS/Carnap-GHCJS.nix { };
-            Carnap-Server = overrideCabal
-                              (oldpkgs.callPackage ./Carnap-Server/Carnap-Server.nix { })
-                              (old: {
-                                preConfigure = ''
-                                  echo "symlinking js in $(pwd)"
-                                  ln -sf ${client-ghcjs.out}/bin/AllActions.jsexe/all.js static/ghcjs/allactions/
-                                  ln -sf ${client-ghcjs.out}/bin/AllActions.jsexe/out.js static/ghcjs/allactions/
-                                  ln -sf ${client-ghcjs.out}/bin/AllActions.jsexe/lib.js static/ghcjs/allactions/
-                                  ln -sf ${client-ghcjs.out}/bin/AllActions.jsexe/runmain.js static/ghcjs/allactions/
-                                  '';
-                                extraLibraries = [ client-ghcjs ];
-                                # Carnap-Server has no tests/they are broken
-                                doCheck = false;
-                                # remove once updated past ghc865
-                                # https://github.com/haskell/haddock/issues/979
-                                doHaddock = false;
-                              });
+            # Carnap-GHCJS  = withGitignore
+            #    (oldpkgs.callPackage ./Carnap-GHCJS/Carnap-GHCJS.nix { });
+
+            Carnap-Server = justStaticExecutables (withGitignore ((overrideCabal
+              (oldpkgs.callPackage ./Carnap-Server/Carnap-Server.nix { })
+              (old: let book = ./Carnap-Book; in {
+                preConfigure = ''
+                  mkdir -p $out/share
+                  echo ":: Copying Carnap-Server data"
+                  cp -r ${book} $out/share/book
+
+                  echo ":: Copying js in $(pwd)"
+                  find static/ghcjs/allactions/ -type l -delete
+                  cp ${client-ghcjs.out}/bin/AllActions.jsexe/all.js static/ghcjs/allactions/
+                  cp ${client-ghcjs.out}/bin/AllActions.jsexe/out.js static/ghcjs/allactions/
+                  cp ${client-ghcjs.out}/bin/AllActions.jsexe/lib.js static/ghcjs/allactions/
+                  cp ${client-ghcjs.out}/bin/AllActions.jsexe/runmain.js static/ghcjs/allactions/
+                  echo ":: Adding a universal settings file"
+                  cp config/settings-example.yml config/settings.yml
```

Comment:
> Will this overwrite an existing settings file? Perhaps add `-n`?

Nope, it is running in a build directory with a copy of the source tree after gitignore is applied. It will override the settings.yml, yes, but it would also do that with `-n` due to the gitignore being applied to the source tree. Regardless, this bit is only used for packaging and docker for which it is reasonable (for closure size/image size) to just copy the ghcjs stuff we use so we don't accidentally pull in node &etc for the executable outputs as we would if we symlinked.

`make run` is still ok for dev purposes.

Reply:
OK, thanks, I understand what's going on a bit better now.